### ISSUE TITLE
react/no-this-in-sfc: eslint-plugin-react 7.25.0 interprets an old legacy jQuery plugin as a sfc

### ISSUE THREAD

**Issue by doberkofler** [yannickcr/eslint-plugin-react#3054]
I get lot's of false positive `react/no-this-in-sfc` warnings in Typescript (*.ts) Files.
It seems as if TypeScript files with ts (not just tsx) extension would be regarded as react files and of type `$.fn.<lowercase>... = function (): <type> | null` would be interpreted as an `sfc`:

```ts
$.fn.getValueAsStringWeak = function (): string | null {
	const val = this.length === 1 ? this.val() : null;

	return typeof val === 'string' ? val : null;
};
```

**Comment by ljharb** [comment]
Does this not happen in v7.24.0?

**Comment by doberkofler** [comment]
> Does this not happen in v7.24.0?

Yes

**Comment by ljharb** [comment]
I’d expect this to not be identified as a component since it’s not a PascalCase-named function.

**Comment by ljharb** [comment]
Using git bisect, this appears to have been introduced by 9a8a4b59f047a8b168a9050d3f7b361ce8121185 / #2992 / @Wesitos.

**Comment by Wesitos** [comment]
I found this equivalent false positive. I'm debugging this. Not sure how it broke.
```js
obj.notAComponent = function () {
  return this.a || null;
};`
```

**Comment by ljharb** [comment]
I’ve got a local test case as well i can tack on if you make a PR :-)

**Comment by Wesitos** [comment]
I just noticed that I also modified getStatelessComponent. Well... that explains it :b.
